proposals/0000-type-indexed-typeable.rst

```diff
@@ -0,0 +1,488 @@
+a type representation,
+
+.. code-block:: haskell
+
+    -- | The kind of a type.
+    typeRepKind :: TypeRep (a :: k) -> TypeRep k
+
+With a ``Typeable`` constraint we can get a ``TypeRep`` for a (non-kind
+polymorphic) type ``a`` with ``typeRep``,
+
+.. code-block:: haskell
+
+    class Typeable (a :: k)
+
+    typeRep :: forall (a :: k). Typeable a => TypeRep a
+
+Note how, in contrast to ``Data.Typeable.typeRep``, we needn't provide a
+``Proxy`` to ``typeRep``; the desired type propagates through ``TypeRep``\'s
+index.
+
+
+We can pattern match on the structure of a ``TypeRep``. For instance, on type
+constructors,
+
+.. code-block:: haskell
+
+    -- | A type constructor type. This must be a unidirectional pattern since
+    -- we have no way of preventing the user from instantiating a TyCon in an
+    -- ill-kinded manner.
+    pattern TRCon :: forall k (a :: k). TyCon -> TypeRep a
+
+    -- | Information about a type constructor. No means of constructing 'TyCon's
+    -- is provided; the only values of this type available are those from
+    -- 'TypeRep's.
+    data TyCon
+    tyConPackage :: TyCon -> String
+    tyConModule :: TyCon -> String
+    tyConName :: TyCon -> String
+
+Type application can also be decomposed,
+
+.. code-block:: haskell
+
+    -- | A representation of a type application, @a b@. This is a bidirectional pattern.
+    pattern TRApp :: forall k2 (fun :: k2). ()
+                  => forall k1 (a :: k1 -> k2) (b :: k1). (fun ~ a b)
+                  => TypeRep a -> TypeRep b -> TypeRep fun
+
+We can also decompose function types (e.g. ``Int -> String``) in their argument
+(e.g. ``Int``) and result types (``String``). Strictly speaking this can be
+expressed in terms of ``TRApp`` but it seems like a common enough pattern that
+it's worth providing a pattern for it,
+
+.. code-block:: haskell
+
+    pattern TRFun :: forall fun. ()
+                  => forall arg res. (fun ~ (arg -> res))
+                  => TypeRep arg
+                  -> TypeRep res
+                  -> TypeRep fun
+
+We can also test for type equality,
+
+.. code-block:: haskell
+
+    -- | Kind-homogeneous type equality
+    eqTypeRep  :: forall k (a :: k) (b :: k).
+                  TypeRep a -> TypeRep b -> Maybe (a :~: b)
+
+    -- | Kind-heterogeneous type equality
+    eqTypeRep' :: forall k1 k2 (a :: k1) (b :: k2).
+                  TypeRep a -> TypeRep b -> Maybe (a :~~: b)
+
+    -- | Kind-heterogeneous type equality
+    data (a :: k1) :~~: (b :: k2) where
+        HRefl :: a :~~: a
+
+Since ``TypeRep`` is a singleton, we can provide a means of satisfying a
+``Typeable`` constraint with a ``TypeRep`` without loss of coherence,
+
+.. code-block:: haskell
+
+    withTypeable :: TypeRep a -> (Typeable a => b) -> b
+
+We will also see later that it is helpful to quantify over the type index. For
+this we introduce,
+
+
+.. code-block:: haskell
+
+    data SomeTypeRep where
+        SomeTypeRep :: forall a. TypeRep a -> SomeTypeRep
+
```

Comment:
It might indeed.

That said, now that we have `TypeInType`, would it be that crazy to rename `SomeTypeRep` to `Type` (or `*` in old fashioned language)? At the term level, we should just be able to write functions that manipulate runtime representations of types! ([Here's a half-baked exploration of this idea](https://gist.github.com/adamgundry/5448c6e5368bd98583d9594f5b93e47a) using the existing (GHC 8.0.1) `Typeable`.)